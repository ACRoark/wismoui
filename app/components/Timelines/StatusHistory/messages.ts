import { defineMessages } from 'react-intl';

export const scope = 'wismo.components.status-history';

export default defineMessages({
  AWAITING_BILLING: {
    id: `${scope}.AWAITING_BILLING`,
  },
  CANCELED: {
    id: `${scope}.CANCELED`,
  },
  COMPLETED: {
    id: `${scope}.COMPLETED`,
  },
  NEEDS_VERIFICATION: {
    id: `${scope}.NEEDS_VERIFICATION`,
  },
  PROCESSED: {
    id: `${scope}.PROCESSED`,
  },
  REQUESTED: {
    id: `${scope}.REQUESTED`,
  },
  SCHEDULED: {
    id: `${scope}.SCHEDULED`,
  },
});
