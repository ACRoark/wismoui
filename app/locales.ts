export const DEFAULT_LOCALE = 'en';

const PROD_LOCALES = {
  en: 'English',
};

const TEST_LOCALES = {
  en: 'English',
  es: 'Spanish',
};

export const SUPPORTED_LOCALES = process.env.NODE_ENV === 'production' ? PROD_LOCALES : TEST_LOCALES;
