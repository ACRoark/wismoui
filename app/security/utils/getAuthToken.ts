import getCookie from './getCookie';

const getAuthToken = (): string => getCookie('token');

export default getAuthToken;
