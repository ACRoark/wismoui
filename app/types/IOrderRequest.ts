import IOrderRequestBase from './IOrderRequestBase';
import IOrderRequestStatusUpdate from './IOrderRequestStatusUpdate';
import IServiceOrderEvent from './IServiceOrderEvent';
import IVerificationEvent from './IVerificationEvent';

interface IOrderRequest extends IOrderRequestBase {
  contactPhoneNumber: string;
  orderRequestStatusUpdates: IOrderRequestStatusUpdate[];
  serviceOrderEvents: IServiceOrderEvent[];
  verificationEvents: IVerificationEvent[];
}

export default IOrderRequest;
