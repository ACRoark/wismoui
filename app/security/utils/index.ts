import getAuthToken from './getAuthToken';
import setAuthToken from './setAuthToken';
import userIsAuthenticated from './userIsAuthenticated';

export { getAuthToken, setAuthToken, userIsAuthenticated };
