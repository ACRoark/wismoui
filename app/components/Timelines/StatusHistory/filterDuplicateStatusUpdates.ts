import { IOrderRequestStatusUpdate } from 'types';

// orderRequestStatusUpdates must be sorted before filtered for duplicates

const filterDuplicateStatusUpdates = (
  orderRequestStatusUpdates: IOrderRequestStatusUpdate[],
): IOrderRequestStatusUpdate[] => {
  const cache = {};
  const filteredOrderStatusUpdates: IOrderRequestStatusUpdate[] = [];

  for (let i = orderRequestStatusUpdates.length - 1; i >= 0; i--) {
    if (!cache[orderRequestStatusUpdates[i].orderRequestStatus]) {
      cache[orderRequestStatusUpdates[i].orderRequestStatus] = true;
      filteredOrderStatusUpdates.push(orderRequestStatusUpdates[i]);
    }
  }

  return filteredOrderStatusUpdates.reverse();
};

export default filterDuplicateStatusUpdates;
