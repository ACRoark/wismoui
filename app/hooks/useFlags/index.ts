import { useSelector } from 'react-redux';

import { IApplicationRootState, IFeatureFlags } from 'types';

const useFlags = (): IFeatureFlags => useSelector((state: IApplicationRootState): IFeatureFlags => state.config.flags);

export default useFlags;
