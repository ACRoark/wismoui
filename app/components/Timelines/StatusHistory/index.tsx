import Timeline from 'antd/lib/timeline';
import moment from 'moment';
import React, { ReactElement } from 'react';
import { FormattedMessage } from 'react-intl';

import DateStamp from 'components/DateStamp';
import IOrderRequest from 'types/IOrderRequest';
import IOrderRequestStatusUpdate from 'types/IOrderRequestStatusUpdate';
import getServiceDate from 'utils/getServiceDate';
import sortOrderRequestUpdates from 'utils/sortOrderRequestUpdates';

import filterDuplicateStatusUpdates from './filterDuplicateStatusUpdates';
import messages from './messages';

import '../index.less';

interface IStatusHistoryProps {
  order: IOrderRequest;
  orderCreatedDate: string;
}

const StatusHistory: React.FC<IStatusHistoryProps> = (props: IStatusHistoryProps): React.ReactElement => {
  const {
    order,
    order: { orderRequestStatusUpdates },
    orderCreatedDate,
  } = props;

  // add a REQUESTED status, and if missing add a SCHEDULED status
  const orderRequestedUpdate: IOrderRequestStatusUpdate = {
    orderRequestStatus: 'REQUESTED',
    createdAtDate: moment(orderCreatedDate).format('YYYYMMDD'),
    createdAtTime: moment(orderCreatedDate).format('HHmmss'),
  };
  const prependedOrderRequestStatusUpdates = [orderRequestedUpdate, ...orderRequestStatusUpdates];
  const hasScheduledStatus = orderRequestStatusUpdates.find(
    (update: IOrderRequestStatusUpdate): boolean => update.orderRequestStatus === 'SCHEDULED',
  );
  const serviceDate = getServiceDate(order) || orderCreatedDate;
  const orderScheduledUpdate: IOrderRequestStatusUpdate | undefined = !hasScheduledStatus
    ? {
        orderRequestStatus: 'SCHEDULED',
        createdAtDate: moment(serviceDate).format('YYYYMMDD'),
        createdAtTime: '000000',
      }
    : undefined;
  const updatedCompletedDate = prependedOrderRequestStatusUpdates.map(
    (update: IOrderRequestStatusUpdate): IOrderRequestStatusUpdate => {
      if (update.orderRequestStatus === 'AWAITING_BILLING') {
        return {
          createdAtDate: serviceDate,
          createdAtTime: '000000',
          orderRequestStatus: 'AWAITING_BILLING',
        };
      }

      return update;
    },
  );
  const updatedStatuses = orderScheduledUpdate
    ? sortOrderRequestUpdates([orderScheduledUpdate, ...updatedCompletedDate])
    : updatedCompletedDate;

  return (
    <Timeline className="dte-wismo-timeline">
      {filterDuplicateStatusUpdates(updatedStatuses).map(
        (update: IOrderRequestStatusUpdate): ReactElement => (
          <Timeline.Item color="gray" key={update.orderRequestStatus}>
            <FormattedMessage
              {...messages[update.orderRequestStatus]}
              values={{ window: <DateStamp value={update.createdAtDate} /> }}
            />
          </Timeline.Item>
        ),
      )}
    </Timeline>
  );
};

export default StatusHistory;
