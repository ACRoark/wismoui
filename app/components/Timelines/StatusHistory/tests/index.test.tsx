import React from 'react';
import { Provider } from 'react-redux';

import { SUPPORTED_LOCALES } from 'locales';
import { createMockStore, createSnapshotWithIntl } from 'testing/utils';
import IOrderRequest from 'types/IOrderRequest';

import useFlags from 'hooks/useFlags';
import StatusHistory from '..';

import { cancelledServiceRequest, retroStartServiceRequest, startServiceRequest, stopServiceRequest } from './data';

jest.mock('hooks/useFlags');

const mockFlags = useFlags as jest.MockedFunction<typeof useFlags>;
const store = createMockStore();

mockFlags.mockReturnValue({
  canChangeLanguage: true,
  developerMode: true,
});

const expectComponentToMatchSnapshot = (order: IOrderRequest, orderCreatedDate: string, language: string): void => {
  const tree = createSnapshotWithIntl(<Provider store={store}><StatusHistory order={order} orderCreatedDate={orderCreatedDate} /></Provider>, language);

  expect(tree).toMatchSnapshot();
};

describe('StatusHistory', (): void => {
  const languages = Object.keys(SUPPORTED_LOCALES);
  const orderCreatedDate = '2020-07-13T20:57:41.7710501Z';

  languages.forEach((language: string): void => {
    it(`should show status history when language is ${language} and service request is cancelled`, (): void =>
      expectComponentToMatchSnapshot(cancelledServiceRequest, orderCreatedDate, language));

    it(`should show status history when language is ${language} and service request is Start (retroactive order)`, (): void =>
      expectComponentToMatchSnapshot(retroStartServiceRequest, '2020-04-21', language));

    it(`should show status history when language is ${language} and service request is Start`, (): void =>
      expectComponentToMatchSnapshot(startServiceRequest, orderCreatedDate, language));

    it(`should show status history when language is ${language} and service request is Stop`, (): void =>
      expectComponentToMatchSnapshot(stopServiceRequest, orderCreatedDate, language));
  });
});
