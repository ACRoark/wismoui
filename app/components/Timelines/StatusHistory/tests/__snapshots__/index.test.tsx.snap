// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatusHistory should show status history when language is en and service request is Start (retroactive order) 1`] = `
<ul
  className="ant-timeline dte-wismo-timeline"
>
  <li
    className="ant-timeline-item"
  >
    <div
      className="ant-timeline-item-tail"
    />
    <div
      className="ant-timeline-item-head ant-timeline-item-head-gray"
      style={
        Object {
          "borderColor": undefined,
        }
      }
    />
    <div
      className="ant-timeline-item-content"
    >
      Order Received 
      04/21/20
    </div>
  </li>
  <li
    className="ant-timeline-item"
  >
    <div
      className="ant-timeline-item-tail"
    />
    <div
      className="ant-timeline-item-head ant-timeline-item-head-gray"
      style={
        Object {
          "borderColor": undefined,
        }
      }
    />
    <div
      className="ant-timeline-item-content"
    >
      Scheduled 
      04/22/20
    </div>
  </li>
  <li
    className="ant-timeline-item ant-timeline-item-last"
  >
    <div
      className="ant-timeline-item-tail"
    />
    <div
      className="ant-timeline-item-head ant-timeline-item-head-gray"
      style={
        Object {
          "borderColor": undefined,
        }
      }
    />
    <div
      className="ant-timeline-item-content"
    >
      Completed 
      04/22/20
    </div>
  </li>
</ul>
`;

exports[`StatusHistory should show status history when language is en and service request is Start 1`] = `
<ul
  className="ant-timeline dte-wismo-timeline"
>
  <li
    className="ant-timeline-item"
  >
    <div
      className="ant-timeline-item-tail"
    />
    <div
      className="ant-timeline-item-head ant-timeline-item-head-gray"
      style={
        Object {
          "borderColor": undefined,
        }
      }
    />
    <div
      className="ant-timeline-item-content"
    >
      Order Received 
      07/13/20
    </div>
  </li>
  <li
    className="ant-timeline-item"
  >
    <div
      className="ant-timeline-item-tail"
    />
    <div
      className="ant-timeline-item-head ant-timeline-item-head-gray"
      style={
        Object {
          "borderColor": undefined,
        }
      }
    />
    <div
      className="ant-timeline-item-content"
    >
      Verified 
      03/20/20
    </div>
  </li>
  <li
    className="ant-timeline-item ant-timeline-item-last"
  >
    <div
      className="ant-timeline-item-tail"
    />
    <div
      className="ant-timeline-item-head ant-timeline-item-head-gray"
      style={
        Object {
          "borderColor": undefined,
        }
      }
    />
    <div
      className="ant-timeline-item-content"
    >
      Scheduled 
      03/21/20
    </div>
  </li>
</ul>
`;

exports[`StatusHistory should show status history when language is en and service request is Stop 1`] = `
<ul
  className="ant-timeline dte-wismo-timeline"
>
  <li
    className="ant-timeline-item"
  >
    <div
      className="ant-timeline-item-tail"
    />
    <div
      className="ant-timeline-item-head ant-timeline-item-head-gray"
      style={
        Object {
          "borderColor": undefined,
        }
      }
    />
    <div
      className="ant-timeline-item-content"
    >
      Order Received 
      07/13/20
    </div>
  </li>
  <li
    className="ant-timeline-item"
  >
    <div
      className="ant-timeline-item-tail"
    />
    <div
      className="ant-timeline-item-head ant-timeline-item-head-gray"
      style={
        Object {
          "borderColor": undefined,
        }
      }
    />
    <div
      className="ant-timeline-item-content"
    >
      Verified 
      03/20/20
    </div>
  </li>
  <li
    className="ant-timeline-item ant-timeline-item-last"
  >
    <div
      className="ant-timeline-item-tail"
    />
    <div
      className="ant-timeline-item-head ant-timeline-item-head-gray"
      style={
        Object {
          "borderColor": undefined,
        }
      }
    />
    <div
      className="ant-timeline-item-content"
    >
      Scheduled 
      03/21/20
    </div>
  </li>
</ul>
`;

exports[`StatusHistory should show status history when language is en and service request is cancelled 1`] = `
<ul
  className="ant-timeline dte-wismo-timeline"
>
  <li
    className="ant-timeline-item"
  >
    <div
      className="ant-timeline-item-tail"
    />
    <div
      className="ant-timeline-item-head ant-timeline-item-head-gray"
      style={
        Object {
          "borderColor": undefined,
        }
      }
    />
    <div
      className="ant-timeline-item-content"
    >
      Order Received 
      07/13/20
    </div>
  </li>
  <li
    className="ant-timeline-item"
  >
    <div
      className="ant-timeline-item-tail"
    />
    <div
      className="ant-timeline-item-head ant-timeline-item-head-gray"
      style={
        Object {
          "borderColor": undefined,
        }
      }
    />
    <div
      className="ant-timeline-item-content"
    >
      Verified 
      03/20/20
    </div>
  </li>
  <li
    className="ant-timeline-item"
  >
    <div
      className="ant-timeline-item-tail"
    />
    <div
      className="ant-timeline-item-head ant-timeline-item-head-gray"
      style={
        Object {
          "borderColor": undefined,
        }
      }
    />
    <div
      className="ant-timeline-item-content"
    >
      Scheduled 
      03/21/20
    </div>
  </li>
  <li
    className="ant-timeline-item ant-timeline-item-last"
  >
    <div
      className="ant-timeline-item-tail"
    />
    <div
      className="ant-timeline-item-head ant-timeline-item-head-gray"
      style={
        Object {
          "borderColor": undefined,
        }
      }
    />
    <div
      className="ant-timeline-item-content"
    >
      Canceled 
      03/22/20
    </div>
  </li>
</ul>
`;

exports[`StatusHistory should show status history when language is es and service request is Start (retroactive order) 1`] = `
<ul
  className="ant-timeline dte-wismo-timeline"
>
  <li
    className="ant-timeline-item"
  >
    <div
      className="ant-timeline-item-tail"
    />
    <div
      className="ant-timeline-item-head ant-timeline-item-head-gray"
      style={
        Object {
          "borderColor": undefined,
        }
      }
    />
    <div
      className="ant-timeline-item-content"
    >
      Orden Recibida 
      21/04/20
    </div>
  </li>
  <li
    className="ant-timeline-item"
  >
    <div
      className="ant-timeline-item-tail"
    />
    <div
      className="ant-timeline-item-head ant-timeline-item-head-gray"
      style={
        Object {
          "borderColor": undefined,
        }
      }
    />
    <div
      className="ant-timeline-item-content"
    >
      Programado 
      22/04/20
    </div>
  </li>
  <li
    className="ant-timeline-item ant-timeline-item-last"
  >
    <div
      className="ant-timeline-item-tail"
    />
    <div
      className="ant-timeline-item-head ant-timeline-item-head-gray"
      style={
        Object {
          "borderColor": undefined,
        }
      }
    />
    <div
      className="ant-timeline-item-content"
    >
      Terminado 
      22/04/20
    </div>
  </li>
</ul>
`;

exports[`StatusHistory should show status history when language is es and service request is Start 1`] = `
<ul
  className="ant-timeline dte-wismo-timeline"
>
  <li
    className="ant-timeline-item"
  >
    <div
      className="ant-timeline-item-tail"
    />
    <div
      className="ant-timeline-item-head ant-timeline-item-head-gray"
      style={
        Object {
          "borderColor": undefined,
        }
      }
    />
    <div
      className="ant-timeline-item-content"
    >
      Orden Recibida 
      13/07/20
    </div>
  </li>
  <li
    className="ant-timeline-item"
  >
    <div
      className="ant-timeline-item-tail"
    />
    <div
      className="ant-timeline-item-head ant-timeline-item-head-gray"
      style={
        Object {
          "borderColor": undefined,
        }
      }
    />
    <div
      className="ant-timeline-item-content"
    >
      Verificado 
      20/03/20
    </div>
  </li>
  <li
    className="ant-timeline-item ant-timeline-item-last"
  >
    <div
      className="ant-timeline-item-tail"
    />
    <div
      className="ant-timeline-item-head ant-timeline-item-head-gray"
      style={
        Object {
          "borderColor": undefined,
        }
      }
    />
    <div
      className="ant-timeline-item-content"
    >
      Programado 
      21/03/20
    </div>
  </li>
</ul>
`;

exports[`StatusHistory should show status history when language is es and service request is Stop 1`] = `
<ul
  className="ant-timeline dte-wismo-timeline"
>
  <li
    className="ant-timeline-item"
  >
    <div
      className="ant-timeline-item-tail"
    />
    <div
      className="ant-timeline-item-head ant-timeline-item-head-gray"
      style={
        Object {
          "borderColor": undefined,
        }
      }
    />
    <div
      className="ant-timeline-item-content"
    >
      Orden Recibida 
      13/07/20
    </div>
  </li>
  <li
    className="ant-timeline-item"
  >
    <div
      className="ant-timeline-item-tail"
    />
    <div
      className="ant-timeline-item-head ant-timeline-item-head-gray"
      style={
        Object {
          "borderColor": undefined,
        }
      }
    />
    <div
      className="ant-timeline-item-content"
    >
      Verificado 
      20/03/20
    </div>
  </li>
  <li
    className="ant-timeline-item ant-timeline-item-last"
  >
    <div
      className="ant-timeline-item-tail"
    />
    <div
      className="ant-timeline-item-head ant-timeline-item-head-gray"
      style={
        Object {
          "borderColor": undefined,
        }
      }
    />
    <div
      className="ant-timeline-item-content"
    >
      Programado 
      21/03/20
    </div>
  </li>
</ul>
`;

exports[`StatusHistory should show status history when language is es and service request is cancelled 1`] = `
<ul
  className="ant-timeline dte-wismo-timeline"
>
  <li
    className="ant-timeline-item"
  >
    <div
      className="ant-timeline-item-tail"
    />
    <div
      className="ant-timeline-item-head ant-timeline-item-head-gray"
      style={
        Object {
          "borderColor": undefined,
        }
      }
    />
    <div
      className="ant-timeline-item-content"
    >
      Orden Recibida 
      13/07/20
    </div>
  </li>
  <li
    className="ant-timeline-item"
  >
    <div
      className="ant-timeline-item-tail"
    />
    <div
      className="ant-timeline-item-head ant-timeline-item-head-gray"
      style={
        Object {
          "borderColor": undefined,
        }
      }
    />
    <div
      className="ant-timeline-item-content"
    >
      Verificado 
      20/03/20
    </div>
  </li>
  <li
    className="ant-timeline-item"
  >
    <div
      className="ant-timeline-item-tail"
    />
    <div
      className="ant-timeline-item-head ant-timeline-item-head-gray"
      style={
        Object {
          "borderColor": undefined,
        }
      }
    />
    <div
      className="ant-timeline-item-content"
    >
      Programado 
      21/03/20
    </div>
  </li>
  <li
    className="ant-timeline-item ant-timeline-item-last"
  >
    <div
      className="ant-timeline-item-tail"
    />
    <div
      className="ant-timeline-item-head ant-timeline-item-head-gray"
      style={
        Object {
          "borderColor": undefined,
        }
      }
    />
    <div
      className="ant-timeline-item-content"
    >
      Cancelado 
      22/03/20
    </div>
  </li>
</ul>
`;
